// netlify-cms-widget-relation/dist/esm/RelationControl.js
import _debounce from "lodash/debounce";
import _last from "lodash/last";
import _isEmpty from "lodash/isEmpty";
import _find from "lodash/find";

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import PropTypes from 'prop-types';
import ImmutablePropTypes from 'react-immutable-proptypes';
import AsyncSelect from 'react-select/lib/Async';
import { List, Map, fromJS } from 'immutable';
import { reactSelectStyles } from 'netlify-cms-ui-default';

function optionToString(option) {
  return option && option.value ? option.value : '';
}

function convertToOption(raw) {
  if (typeof raw === 'string') {
    return {
      label: raw,
      value: raw
    };
  }

  return Map.isMap(raw) ? raw.toJS() : raw;
}

function getSelectedValue(_ref) {
  let value = _ref.value,
      options = _ref.options,
      isMultiple = _ref.isMultiple;

  if (isMultiple) {
    const selectedOptions = List.isList(value) ? value.toJS() : value;

    if (!selectedOptions || !Array.isArray(selectedOptions)) {
      return null;
    }

    return selectedOptions.map(i => options.find(o => o.value === (i.value || i))).filter(Boolean).map(convertToOption);
  } else {
    return _find(options, ['value', value]) || null;
  }
}

export default class RelationControl extends React.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "didInitialSearch", false);

    _defineProperty(this, "handleChange", selectedOption => {
      const _this$props = this.props,
            onChange = _this$props.onChange,
            field = _this$props.field;
      let value;

      if (Array.isArray(selectedOption)) {
        value = selectedOption.map(optionToString);
        onChange(fromJS(value), {
          [field.get('name')]: {
            [field.get('collection')]: {
              [_last(value)]: !_isEmpty(selectedOption) && _last(selectedOption).data
            }
          }
        });
      } else {
        value = optionToString(selectedOption);
        onChange(value, {
          [field.get('name')]: {
            [field.get('collection')]: {
              [value]: selectedOption.data
            }
          }
        });
      }
    });

    _defineProperty(this, "parseHitOptions", hits => {
      const field = this.props.field;
      const valueField = field.get('valueField');
      const displayField = field.get('displayFields') || field.get('valueField');
      return hits.map(hit => {
        return {
          data: hit.data,
          value: hit.data[valueField],
          label: List.isList(displayField) ? displayField.toJS().map(key => hit.data[key]).join(' ') : hit.data[displayField]
        };
      });
    });

    _defineProperty(this, "loadOptions", _debounce((term, callback) => {
      const _this$props2 = this.props,
            field = _this$props2.field,
            query = _this$props2.query,
            forID = _this$props2.forID;
      const collection = field.get('collection');
      const searchFields = field.get('searchFields');
      const searchFieldsArray = List.isList(searchFields) ? searchFields.toJS() : [searchFields];
      query(forID, collection, searchFieldsArray, term).then((_ref2) => {
        let payload = _ref2.payload;
        let options = this.parseHitOptions(payload.response.hits);

        if (!this.allOptions && !term) {
          this.allOptions = options;
        }

        if (!term) {
          options = options.slice(0, 20);
        }

        callback(options);
      });
    }, 500));
  }

  shouldComponentUpdate(nextProps) {
    return this.props.value !== nextProps.value || this.props.hasActiveStyle !== nextProps.hasActiveStyle || this.props.queryHits !== nextProps.queryHits;
  }

  componentDidUpdate(prevProps) {
    /**
     * Load extra post data into the store after first query.
     */
    if (this.didInitialSearch) return;
    const _this$props3 = this.props,
          value = _this$props3.value,
          field = _this$props3.field,
          forID = _this$props3.forID,
          queryHits = _this$props3.queryHits,
          onChange = _this$props3.onChange;

    if (queryHits !== prevProps.queryHits && queryHits.get(forID)) {
      this.didInitialSearch = true;
      const valueField = field.get('valueField');
      const hits = queryHits.get(forID);

      if (value) {
        const listValue = List.isList(value) ? value : List([value]);
        listValue.forEach(val => {
          const hit = hits.find(i => i.data[valueField] === val);

          if (hit) {
            onChange(value, {
              [field.get('name')]: {
                [field.get('collection')]: {
                  [val]: hit.data
                }
              }
            });
          }
        });
      }
    }
  }

  render() {
    const _this$props4 = this.props,
          value = _this$props4.value,
          field = _this$props4.field,
          forID = _this$props4.forID,
          classNameWrapper = _this$props4.classNameWrapper,
          setActiveStyle = _this$props4.setActiveStyle,
          setInactiveStyle = _this$props4.setInactiveStyle,
          queryHits = _this$props4.queryHits;
    const isMultiple = field.get('multiple', false);
    const isClearable = !field.get('required', true) || isMultiple;
    const hits = queryHits.get(forID, []);
    const options = this.allOptions || this.parseHitOptions(hits);
    const selectedValue = getSelectedValue({
      options,
      value,
      isMultiple
    });
    return React.createElement(AsyncSelect, {
      value: selectedValue,
      inputId: forID,
      defaultOptions: true,
      loadOptions: this.loadOptions,
      onChange: this.handleChange,
      className: classNameWrapper,
      onFocus: setActiveStyle,
      onBlur: setInactiveStyle,
      styles: reactSelectStyles,
      isMulti: isMultiple,
      isClearable: isClearable,
      placeholder: ""
    });
  }

}

_defineProperty(RelationControl, "propTypes", {
  onChange: PropTypes.func.isRequired,
  forID: PropTypes.string.isRequired,
  value: PropTypes.node,
  field: ImmutablePropTypes.map,
  fetchID: PropTypes.string,
  query: PropTypes.func.isRequired,
  queryHits: PropTypes.oneOfType([PropTypes.array, PropTypes.object]),
  classNameWrapper: PropTypes.string.isRequired,
  setActiveStyle: PropTypes.func.isRequired,
  setInactiveStyle: PropTypes.func.isRequired
});
