// netlify-cms-widget-markdown/dist/esm/MarkdownControl/renderers.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

/* eslint-disable react/prop-types */
import React from 'react';
import Shortcode from "./Shortcode";
/**
 * Slate uses React components to render each type of node that it receives.
 * This is the closest thing Slate has to a schema definition. The types are set
 * by us when we manually deserialize from Remark's MDAST to Slate's AST.
 */

/**
 * Mark Components
 */

const Bold = props => React.createElement("strong", null, props.children);

const Italic = props => React.createElement("em", null, props.children);

const Strikethrough = props => React.createElement("s", null, props.children);

const Code = props => React.createElement("code", null, props.children);
/**
 * Node Components
 */


const Paragraph = props => React.createElement("p", props.attributes, props.children);

const ListItem = props => React.createElement("li", props.attributes, props.children);

const Quote = props => React.createElement("blockquote", props.attributes, props.children);

const CodeBlock = props => React.createElement("pre", null, React.createElement("code", props.attributes, props.children));

const HeadingOne = props => React.createElement("h1", props.attributes, props.children);

const HeadingTwo = props => React.createElement("h2", props.attributes, props.children);

const HeadingThree = props => React.createElement("h3", props.attributes, props.children);

const HeadingFour = props => React.createElement("h4", props.attributes, props.children);

const HeadingFive = props => React.createElement("h5", props.attributes, props.children);

const HeadingSix = props => React.createElement("h6", props.attributes, props.children);

const Table = props => React.createElement("table", null, React.createElement("tbody", props.attributes, props.children));

const TableRow = props => React.createElement("tr", props.attributes, props.children);

const TableCell = props => React.createElement("td", props.attributes, props.children);

const ThematicBreak = props => React.createElement("hr", props.attributes);

const BulletedList = props => React.createElement("ul", props.attributes, props.children);

const NumberedList = props => React.createElement("ol", _extends({}, props.attributes, {
  start: props.node.data.get('start') || 1
}), props.children);

const Link = props => {
  const data = props.node.get('data');
  const marks = data.get('marks');
  const url = data.get('url');
  const title = data.get('title');
  const link = React.createElement("a", _extends({
    href: url,
    title: title
  }, props.attributes), props.children);
  const result = !marks ? link : marks.reduce((acc, mark) => {
    return renderMark({
      mark,
      children: acc
    });
  }, link);
  return result;
};

const Image = props => {
  const data = props.node.get('data');
  const marks = data.get('marks');
  const url = data.get('url');
  const title = data.get('title');
  const alt = data.get('alt');
  const image = React.createElement("img", _extends({
    src: url,
    title: title,
    alt: alt
  }, props.attributes));
  const result = !marks ? image : marks.reduce((acc, mark) => {
    return renderMark({
      mark,
      children: acc
    });
  }, image);
  return result;
};

export const renderMark = props => {
  switch (props.mark.type) {
    case 'bold':
      return React.createElement(Bold, props);

    case 'italic':
      return React.createElement(Italic, props);

    case 'strikethrough':
      return React.createElement(Strikethrough, props);

    case 'code':
      return React.createElement(Code, props);
  }
};
export const renderNode = props => {
  switch (props.node.type) {
    case 'paragraph':
      return React.createElement(Paragraph, props);

    case 'list-item':
      return React.createElement(ListItem, props);

    case 'quote':
      return React.createElement(Quote, props);

    case 'code':
      return React.createElement(CodeBlock, props);

    case 'heading-one':
      return React.createElement(HeadingOne, props);

    case 'heading-two':
      return React.createElement(HeadingTwo, props);

    case 'heading-three':
      return React.createElement(HeadingThree, props);

    case 'heading-four':
      return React.createElement(HeadingFour, props);

    case 'heading-five':
      return React.createElement(HeadingFive, props);

    case 'heading-six':
      return React.createElement(HeadingSix, props);

    case 'table':
      return React.createElement(Table, props);

    case 'table-row':
      return React.createElement(TableRow, props);

    case 'table-cell':
      return React.createElement(TableCell, props);

    case 'thematic-break':
      return React.createElement(ThematicBreak, props);

    case 'bulleted-list':
      return React.createElement(BulletedList, props);

    case 'numbered-list':
      return React.createElement(NumberedList, props);

    case 'link':
      return React.createElement(Link, props);

    case 'image':
      return React.createElement(Image, props);

    case 'shortcode':
      return React.createElement(Shortcode, props);
  }
};
