// netlify-cms-core/dist/esm/components/MediaLibrary/MediaLibrary.js
import _map from "lodash/map";
import _orderBy from "lodash/orderBy";

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import PropTypes from 'prop-types';
import ImmutablePropTypes from 'react-immutable-proptypes';
import { connect } from 'react-redux';
import { translate } from 'react-polyglot';
import fuzzy from 'fuzzy';
import { resolvePath, fileExtension } from 'netlify-cms-lib-util';
import { loadMedia as loadMediaAction, persistMedia as persistMediaAction, deleteMedia as deleteMediaAction, insertMedia as insertMediaAction, loadMediaDisplayURL as loadMediaDisplayURLAction, closeMediaLibrary as closeMediaLibraryAction } from "../../actions/mediaLibrary";
import MediaLibraryModal from "./MediaLibraryModal";
/**
 * Extensions used to determine which files to show when the media library is
 * accessed from an image insertion field.
 */

const IMAGE_EXTENSIONS_VIEWABLE = ['jpg', 'jpeg', 'webp', 'gif', 'png', 'bmp', 'tiff', 'svg'];
const IMAGE_EXTENSIONS = [...IMAGE_EXTENSIONS_VIEWABLE];
const fileShape = {
  displayURL: PropTypes.oneOfType([PropTypes.string, PropTypes.object]).isRequired,
  id: PropTypes.string.isRequired,
  key: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  queryOrder: PropTypes.number,
  size: PropTypes.number,
  url: PropTypes.string,
  urlIsPublicPath: PropTypes.bool
};

class MediaLibrary extends React.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      selectedFile: {},
      query: ''
    });

    _defineProperty(this, "loadDisplayURL", file => {
      const loadMediaDisplayURL = this.props.loadMediaDisplayURL;
      loadMediaDisplayURL(file);
    });

    _defineProperty(this, "filterImages", files => {
      return files.filter(file => {
        const ext = fileExtension(file.name).toLowerCase();
        return IMAGE_EXTENSIONS.includes(ext);
      });
    });

    _defineProperty(this, "toTableData", files => {
      const tableData = files && files.map((_ref) => {
        let key = _ref.key,
            name = _ref.name,
            id = _ref.id,
            size = _ref.size,
            queryOrder = _ref.queryOrder,
            url = _ref.url,
            urlIsPublicPath = _ref.urlIsPublicPath,
            displayURL = _ref.displayURL;
        const ext = fileExtension(name).toLowerCase();
        return {
          key,
          id,
          name,
          type: ext.toUpperCase(),
          size,
          queryOrder,
          url,
          urlIsPublicPath,
          displayURL,
          isImage: IMAGE_EXTENSIONS.includes(ext),
          isViewableImage: IMAGE_EXTENSIONS_VIEWABLE.includes(ext)
        };
      });
      /**
       * Get the sort order for use with `lodash.orderBy`, and always add the
       * `queryOrder` sort as the lowest priority sort order.
       */

      const sortFields = this.state.sortFields;

      const fieldNames = _map(sortFields, 'fieldName').concat('queryOrder');

      const directions = _map(sortFields, 'direction').concat('asc');

      return _orderBy(tableData, fieldNames, directions);
    });

    _defineProperty(this, "handleClose", () => {
      this.props.closeMediaLibrary();
    });

    _defineProperty(this, "handleAssetClick", asset => {
      const selectedFile = this.state.selectedFile.key === asset.key ? {} : asset;
      this.setState({
        selectedFile
      });
    });

    _defineProperty(this, "handlePersist", async event => {
      /**
       * Stop the browser from automatically handling the file input click, and
       * get the file for upload, and retain the synthetic event for access after
       * the asynchronous persist operation.
       */
      event.stopPropagation();
      event.preventDefault();
      event.persist();
      const _this$props = this.props,
            persistMedia = _this$props.persistMedia,
            privateUpload = _this$props.privateUpload;

      const _ref2 = event.dataTransfer || event.target,
            fileList = _ref2.files;

      const files = [...fileList];
      const file = files[0];
      await persistMedia(file, {
        privateUpload
      });
      event.target.value = null;
      this.scrollToTop();
    });

    _defineProperty(this, "handleInsert", () => {
      const selectedFile = this.state.selectedFile;
      const name = selectedFile.name,
            url = selectedFile.url,
            urlIsPublicPath = selectedFile.urlIsPublicPath;
      const _this$props2 = this.props,
            insertMedia = _this$props2.insertMedia,
            publicFolder = _this$props2.publicFolder;
      const publicPath = urlIsPublicPath ? url : resolvePath(name, publicFolder);
      insertMedia(publicPath);
      this.handleClose();
    });

    _defineProperty(this, "handleDelete", () => {
      const selectedFile = this.state.selectedFile;
      const _this$props3 = this.props,
            files = _this$props3.files,
            deleteMedia = _this$props3.deleteMedia,
            privateUpload = _this$props3.privateUpload,
            t = _this$props3.t;

      if (!window.confirm(t('mediaLibrary.mediaLibrary.onDelete'))) {
        return;
      }

      const file = files.find(file => selectedFile.key === file.key);
      deleteMedia(file, {
        privateUpload
      }).then(() => {
        this.setState({
          selectedFile: {}
        });
      });
    });

    _defineProperty(this, "handleLoadMore", () => {
      const _this$props4 = this.props,
            loadMedia = _this$props4.loadMedia,
            dynamicSearchQuery = _this$props4.dynamicSearchQuery,
            page = _this$props4.page,
            privateUpload = _this$props4.privateUpload;
      loadMedia({
        query: dynamicSearchQuery,
        page: page + 1,
        privateUpload
      });
    });

    _defineProperty(this, "handleSearchKeyDown", async event => {
      const _this$props5 = this.props,
            dynamicSearch = _this$props5.dynamicSearch,
            loadMedia = _this$props5.loadMedia,
            privateUpload = _this$props5.privateUpload;

      if (event.key === 'Enter' && dynamicSearch) {
        await loadMedia({
          query: this.state.query,
          privateUpload
        });
        this.scrollToTop();
      }
    });

    _defineProperty(this, "scrollToTop", () => {
      this.scrollContainerRef.scrollTop = 0;
    });

    _defineProperty(this, "handleSearchChange", event => {
      this.setState({
        query: event.target.value
      });
    });

    _defineProperty(this, "queryFilter", (query, files) => {
      /**
       * Because file names don't have spaces, typing a space eliminates all
       * potential matches, so we strip them all out internally before running the
       * query.
       */
      const strippedQuery = query.replace(/ /g, '');
      const matches = fuzzy.filter(strippedQuery, files, {
        extract: file => file.name
      });
      const matchFiles = matches.map((match, queryIndex) => {
        const file = files[match.index];
        return _objectSpread({}, file, {
          queryIndex
        });
      });
      return matchFiles;
    });
  }

  componentDidMount() {
    this.props.loadMedia();
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    /**
     * We clear old state from the media library when it's being re-opened
     * because, when doing so on close, the state is cleared while the media
     * library is still fading away.
     */
    const isOpening = !this.props.isVisible && nextProps.isVisible;

    if (isOpening) {
      this.setState({
        selectedFile: {},
        query: ''
      });
    }
  }

  componentDidUpdate(prevProps) {
    const isOpening = !prevProps.isVisible && this.props.isVisible;

    if (isOpening && prevProps.privateUpload !== this.props.privateUpload) {
      this.props.loadMedia({
        privateUpload: this.props.privateUpload
      });
    }
  }

  render() {
    const _this$props6 = this.props,
          isVisible = _this$props6.isVisible,
          canInsert = _this$props6.canInsert,
          files = _this$props6.files,
          dynamicSearch = _this$props6.dynamicSearch,
          dynamicSearchActive = _this$props6.dynamicSearchActive,
          forImage = _this$props6.forImage,
          isLoading = _this$props6.isLoading,
          isPersisting = _this$props6.isPersisting,
          isDeleting = _this$props6.isDeleting,
          hasNextPage = _this$props6.hasNextPage,
          isPaginating = _this$props6.isPaginating,
          privateUpload = _this$props6.privateUpload,
          displayURLs = _this$props6.displayURLs,
          t = _this$props6.t;
    return React.createElement(MediaLibraryModal, {
      isVisible: isVisible,
      canInsert: canInsert,
      files: files,
      dynamicSearch: dynamicSearch,
      dynamicSearchActive: dynamicSearchActive,
      forImage: forImage,
      isLoading: isLoading,
      isPersisting: isPersisting,
      isDeleting: isDeleting,
      hasNextPage: hasNextPage,
      isPaginating: isPaginating,
      privateUpload: privateUpload,
      query: this.state.query,
      selectedFile: this.state.selectedFile,
      handleFilter: this.filterImages,
      handleQuery: this.queryFilter,
      toTableData: this.toTableData,
      handleClose: this.handleClose,
      handleSearchChange: this.handleSearchChange,
      handleSearchKeyDown: this.handleSearchKeyDown,
      handlePersist: this.handlePersist,
      handleDelete: this.handleDelete,
      handleInsert: this.handleInsert,
      setScrollContainerRef: ref => this.scrollContainerRef = ref,
      handleAssetClick: this.handleAssetClick,
      handleLoadMore: this.handleLoadMore,
      displayURLs: displayURLs,
      loadDisplayURL: this.loadDisplayURL,
      t: t
    });
  }

}

_defineProperty(MediaLibrary, "propTypes", {
  isVisible: PropTypes.bool,
  loadMediaDisplayURL: PropTypes.func,
  displayURLs: ImmutablePropTypes.map,
  canInsert: PropTypes.bool,
  files: PropTypes.arrayOf(PropTypes.shape(fileShape)).isRequired,
  dynamicSearch: PropTypes.bool,
  dynamicSearchActive: PropTypes.bool,
  forImage: PropTypes.bool,
  isLoading: PropTypes.bool,
  isPersisting: PropTypes.bool,
  isDeleting: PropTypes.bool,
  hasNextPage: PropTypes.bool,
  isPaginating: PropTypes.bool,
  privateUpload: PropTypes.bool,
  loadMedia: PropTypes.func.isRequired,
  dynamicSearchQuery: PropTypes.string,
  page: PropTypes.number,
  persistMedia: PropTypes.func.isRequired,
  deleteMedia: PropTypes.func.isRequired,
  insertMedia: PropTypes.func.isRequired,
  publicFolder: PropTypes.string,
  closeMediaLibrary: PropTypes.func.isRequired,
  t: PropTypes.func.isRequired
});

_defineProperty(MediaLibrary, "defaultProps", {
  files: []
});

const mapStateToProps = state => {
  const config = state.config,
        mediaLibrary = state.mediaLibrary;
  const configProps = {
    publicFolder: config.get('public_folder')
  };
  const mediaLibraryProps = {
    isVisible: mediaLibrary.get('isVisible'),
    canInsert: mediaLibrary.get('canInsert'),
    files: mediaLibrary.get('files'),
    displayURLs: mediaLibrary.get('displayURLs'),
    dynamicSearch: mediaLibrary.get('dynamicSearch'),
    dynamicSearchActive: mediaLibrary.get('dynamicSearchActive'),
    dynamicSearchQuery: mediaLibrary.get('dynamicSearchQuery'),
    forImage: mediaLibrary.get('forImage'),
    isLoading: mediaLibrary.get('isLoading'),
    isPersisting: mediaLibrary.get('isPersisting'),
    isDeleting: mediaLibrary.get('isDeleting'),
    privateUpload: mediaLibrary.get('privateUpload'),
    page: mediaLibrary.get('page'),
    hasNextPage: mediaLibrary.get('hasNextPage'),
    isPaginating: mediaLibrary.get('isPaginating')
  };
  return _objectSpread({}, configProps, mediaLibraryProps);
};

const mapDispatchToProps = {
  loadMedia: loadMediaAction,
  persistMedia: persistMediaAction,
  deleteMedia: deleteMediaAction,
  insertMedia: insertMediaAction,
  loadMediaDisplayURL: loadMediaDisplayURLAction,
  closeMediaLibrary: closeMediaLibraryAction
};
export default connect(mapStateToProps, mapDispatchToProps)(translate()(MediaLibrary));
